### PR TITLE
Improved performance and dtype bug fixes for FermionOperator2nd.

### PR DESCRIPTION
**Pull request by jwnys**
Improved performance:
- max_conn_size is typically lower now, since we count only one connected element per hermitian conjugate pair

Bug fixes:
- Copy did not have the correct dtype due to inplace changes
- Double terms did not cancel

Added a few extra tests.

### CODE REVIEW

File: netket/experimental/operator/_fermions_2nd.py [hunk 1]
```diff
@@ -217,16 +225,44 @@ def copy(self, *, dtype: Optional[DType] = None):
...
+                mcs += n_non_herm_terms // 2  # only half
+            else:
+                ops = _reduce_operators(self._operators, self.dtype)
+                mcs += len(ops)
+            self._max_conn_size = mcs
+        return self._max_conn_size
+        # if _isclose(self._constant, 0):
+        #     return len(self._operators)
+        # else:
+        #     return len(self._operators) + 1  # constant also can add a term
```
Comment (PhilipVinc) [inline netket/experimental/operator/_fermions_2nd.py:265]:
```suggestion
```

File: netket/experimental/operator/_fermions_2nd.py [hunk 2]
```diff
@@ -603,7 +639,14 @@ def _canonicalize_input(terms, weights, constant, dtype):
...
 
-    operators = dict(zip(terms, weights))
+    # add the weights of terms that occur multiple times
+    def dtype_init():
+        return np.array(0, dtype=dtype)
+
+    operators = defaultdict(dtype_init)
+    for t, w in zip(terms, weights):
+        operators[t] += w
+    operators = _remove_dict_zeros(dict(operators))
```
Comment (PhilipVinc) [inline netket/experimental/operator/_fermions_2nd.py:649]: what is this doing? I don't understand it.

File: netket/experimental/operator/_fermions_2nd.py [hunk 3]
```diff
@@ -217,16 +225,44 @@ def copy(self, *, dtype: Optional[DType] = None):
...
+        terms = self._operators.keys()
+        weights = np.array(list(self._operators.values()), dtype=dtype)
+        op._operators = dict(zip(terms, weights))
         return op
 
+    def reduce(self):
+        """Reduce the number of operators by bringing them to normal form and grouping equivalent terms"""
+        self._operators = _reduce_operators(self._operators, self.dtype)
+        return self
+
```
Comment (PhilipVinc) [inline netket/experimental/operator/_fermions_2nd.py:238]: We should decide if this should be an inplace operation as you made it here (aka, which modifies the existing operator) or if it should be out of place and it does not modify it but returns a new operator which is _reduced_.

In general, out-of-place non modifying APIs are easier to reason and introduce less bugs (especially if we wrap operators inside operators) so I'd lean towards that, but would like to have opinions.

Comment (gcarleo) [reply netket/experimental/operator/_fermions_2nd.py:238]: yes I agree out of place is easier to debug, however maybe Jannes here means to do a default simplification (i.e. this `reduce` should be called on every instance of the class?) in this case then maybe I would not expose `reduce` as a public function and just make it internal, non returning?

**Comment by jwnys** [comment]
Btw, #1279 is a more straightforward optimization of the diagonal operators and is closer to being ready for deployment. The branch here is only a draft and WIP.

**Comment by PhilipVinc** [comment]
please rebase on current master in order to fix all the CI errors.

**Comment by PhilipVinc** [comment]
I think all the changes from this PR made it up into netket so I will close. Feel free to bring it up if I'm missing something.

**Pull request closed without merging**